Added Multi-window Telemetry

keianhzo opened the pull request:
Added telemetry events from #1329

MortimerGoro commented:
```
/opt/FirefoxReality/app/src/main/res/layout/context_menu_item.xml:15: AAPT: error: resource drawable/context_menu_item_background_color (aka org.mozilla.vrbrowser:drawable/context_menu_item_background_color) not found.
error: failed linking file resources.
```

daoshengmu commented on app/src/common/shared/org/mozilla/vrbrowser/VRBrowserActivity.java:
```diff
@@ -344,6 +348,8 @@ protected void onResume() {
     @Override
     protected void onDestroy() {
+        TelemetryWrapper.destroy();
+
```
I am curious why you want to queue your multiple window telemetry at `onDestory()` instead of be like our previous telemetry, queue them at `onStart()` and `onStop()`?

keianhzo replied:
Some windows related events like window lifetime histogram or window placement pct are better measured at the end of the session as you can stop and resume but the windows are not recreated so the stat wouldn't be completely accurate if we flush the data in the middle. We could artificially trigger a window destroy/re-create at onStop but this just made things simpler.

daoshengmu replied:
put in onDestroy, I would be afraid the metric will not be uploaded for Oculus release. We usually do it `onStart/onStop` for `recordSessionStart/recordSessionEnd` and then scheduleUpload().

IIUC, `onDestroy ` will be called after `onStop`?

keianhzo replied:
Fixed this

keianhzo commented:
@daoshengmu I've added some logs so all the multi-window related telemetry can be seen at the end of the session. Let me know if you catch anything.

daoshengmu commented:
We probably will start to replace our existing telemetry service to `Glean` in Q3, because the telemetry service will not be maintained soon. We need to keep an eye on it for the coming changes.

MortimerGoro commented on app/src/common/shared/org/mozilla/vrbrowser/telemetry/TelemetryWrapper.java:
```diff
@@ -292,5 +344,217 @@ public static void uploadImmersiveToHistogram() {
+        }
+    }
+
+    private static void queueOpenWindowsAvgEvent() {
```
Should we measure time here too (or in a extra data) as you are doing in `queueActiveWindowPctEvent`? You can have 1.5 windows open in a session but how do you know how long they have used 1 windows or 2 windows?

keianhzo replied:
I think the goal of that stat is knowing what's the average of open windows in a session but we can also make a a % of how much time the user spends with 1, 2 or 3 open windows. It just depends of what's more useful for us.

jvonitter replied:
% of how much time the user spends with 1, 2 or 3 open windows would be great! probably more useful than the average number.

keianhzo replied:
@jvonitter ok, I'll add both :)

keianhzo commented:
@daoshengmu Just FYI I've removed the the Telemetry queueing from `start` as it’s always empty at that point (we always flush in on `stop`) so it doesn’t make sense to queue it. I've just removed the data queuing NOT the upload scheduling, we still need that to solve the boot permission issues.

daoshengmu commented:
> @daoshengmu Just FYI I've removed the the Telemetry queueing from `start` as it’s always empty at that point (we always flush in on `stop`) so it doesn’t make sense to queue it. I've just removed the data queuing NOT the upload scheduling, we still need that to solve the boot permission issues.

I just took a look at `queueHistogram()`, it will write file stream into a storage. Every bucket is 0 value from `start`, so it will not affect the final result. But, they are still redundant data.

According to https://github.com/MozillaReality/FirefoxReality/issues/1353#issuecomment-508170240, our `BOOT` permission will only cause the regression of the job scheduler. So, I agree we can remove the `queueHistogram()` at `start`. Thanks.
